### ISSUE TITLE
Do not include user while checking bypass cookie if not bypass_v2

### ISSUE THREAD

**Issue by sonertari** [e2guardian/e2guardian#520]
Otherwise bypass cookie does not work without user configuration. The following is my proposal to fix the issue, which works here. What this patch is doing is actually the same thing as in ConnectionHandler::hashedCookie() function. But we could pass the `FOptionContainer*` var instead, as the magic param is also in the same container. As you wish.
```
diff -ruN e2guardian-5.3.2.orig/src/ConnectionHandler.cpp e2guardian-5.3.2/src/ConnectionHandler.cpp
--- e2guardian-5.3.2.orig/src/ConnectionHandler.cpp	2019-03-15 13:39:09.000000000 +0300
+++ e2guardian-5.3.2/src/ConnectionHandler.cpp	2019-03-16 21:41:28.631563000 +0300
@@ -2924,7 +2924,7 @@
             }
         } else if (ldl->fg[filtergroup]->bypass_mode != 0) {
             if (header.isBypassCookie(checkme.urldomain, ldl->fg[filtergroup]->cookie_magic.c_str(),
-                                      clientip.c_str(), clientuser.c_str())) {
+                                      clientip.c_str(), clientuser.c_str(), ldl->fg[filtergroup]->bypass_v2)) {
 #ifdef DGDEBUG
                 std::cerr << thread_id << " -Bypass cookie match" << std::endl;
 #endif
diff -ruN e2guardian-5.3.2.orig/src/HTTPHeader.cpp e2guardian-5.3.2/src/HTTPHeader.cpp
--- e2guardian-5.3.2.orig/src/HTTPHeader.cpp	2019-03-15 13:39:09.000000000 +0300
+++ e2guardian-5.3.2/src/HTTPHeader.cpp	2019-03-16 22:13:28.433000413 +0300
@@ -1312,7 +1312,7 @@
 }
 
 // is this a temporary filter bypass cookie?
-bool HTTPHeader::isBypassCookie(String url, const char *magic, const char *clientip, const char *user)
+bool HTTPHeader::isBypassCookie(String url, const char *magic, const char *clientip, const char *user, const bool bypass_v2)
 {
     String cookie(getCookie("GBYPASS"));
     if (!cookie.length()) {
@@ -1325,7 +1325,8 @@
     String cookietime(cookie.after(cookiehash.toCharArray()));
     String mymagic(magic);
     mymagic += clientip;
-    mymagic += user;
+    if (bypass_v2)
+        mymagic += user;
     mymagic += cookietime;
     bool matched = false;
     while (url.contains(".")) {
diff -ruN e2guardian-5.3.2.orig/src/HTTPHeader.hpp e2guardian-5.3.2/src/HTTPHeader.hpp
--- e2guardian-5.3.2.orig/src/HTTPHeader.hpp	2019-03-15 13:39:09.000000000 +0300
+++ e2guardian-5.3.2/src/HTTPHeader.hpp	2019-03-16 21:42:32.548657000 +0300
@@ -170,7 +170,7 @@
 
     // Bypass URL & Cookie funcs
 
-    bool isBypassCookie(String url, const char *magic, const char *clientip, const char *user);
+    bool isBypassCookie(String url, const char *magic, const char *clientip, const char *user, const bool bypass_v2);
     void chopBypass(String url, bool infectionbypass);
     void chopScanBypass(String url);
     // add cookie to outgoing headers with given name & value
```

**Comment by fredbcode** [comment]
Thank you, what happen exactly? I'm using bypass without any issue with 5.3.2

**Comment by sonertari** [comment]
My report does not need further explanation.

**Comment by HadiJadallah** [comment]
Hello @fredbcode.

Maybe the code that generates the hashedCookie is checking for bypass_v2 to add the clientuser to the hash while when checking in isBypassCookie it does not??

not sure.

String ConnectionHandler::hashedCookie(String *url, const char *magic, std::string *clientip, int bypasstimestamp) {
........
    **if(ldl->fg[filtergroup]->bypass_v2)
            data += clientuser;**
    data += timecode;
......
}

and in 
bool HTTPHeader::isBypassCookie(String url, const char *magic, const char *clientip, const char *user)
{
    String cookie(getCookie("GBYPASS"));
    if (!cookie.length()) {
         ......
        return false;
    }
   .......

   **mymagic += user;**

    .........
}

**Comment by HadiJadallah** [comment]
Hi,

Should'nt isBypassCookie check if bypass_v2 is set before adding user info to mymagic?
Per the hash generating routine for the cookie, the code do check if bypass_v2 is set before adding the user info to the hashcode.

not sure since I never used bypass version 1.

**Comment by sonertari** [comment]
@HadiJadallah, exactly.

**Comment by philipianpearce** [comment]
This issue proved not to be to do with the bypass version, but with the mismatch between cookie generation and cookie detection.   By-pass cookies are only used internally by e2g and so there should be no difference between v1 and v2 for them.   Odd that no-one picked this up before as the bug has been around for some time.   Thanks for logging it as an issue!

Fixed in v5.3.dev (so will be in v5.3.5), v5.4.dev and v5.5.dev.

Philip
